File: Data Structures & Algorithms/three-integer-sum/submission-0.go
import "slices"

func threeSum(nums []int) [][]int {
	slices.Sort(nums)
	res := [][]int{}

	for i := 0; i < len(nums)-2; i++ {
		if nums[i] > 0 {
			break
		}

		if i > 0 && nums[i] == nums[i-1] {
			continue
		}

		left := i + 1
		right := len(nums) - 1
		target := -nums[i]

		for left < right {
			sum := nums[left] + nums[right]

			if sum == target {
				res = append(res, []int{nums[i], nums[left], nums[right]})
				left++
				right--

				for left < right && nums[left] == nums[left-1] {
					left++
				}
				for left < right && nums[right] == nums[right+1] {
					right--
				}
			} else if sum < target {
				left++
			} else {
				right--
			}
		}
	}

	return res
}

File: Data Structures & Algorithms/search-2d-matrix/submission-0.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        for j in range(len(matrix)): 
            resR = self.search(matrix[j], target) 
            if resR == -1: 
                continue
            else: 
                return True
                
        return False
        
    def search(self, nums: List[int], target: int) -> int:
        left = 0 
        right = len(nums) - 1
        
        while left <= right: 
            mid = (left + right) // 2
            
            if nums[mid] == target:
                return mid
                
            if nums[mid] < target: 
                left = mid + 1 
            else: 
                right = mid - 1 
                
        return -1 
            